diff --git a/node_modules/@koinos/sdk-as/assembly/util/space.ts b/node_modules/@koinos/sdk-as/assembly/util/space.ts
index 5c44c4f..3fcc7ef 100644
--- a/node_modules/@koinos/sdk-as/assembly/util/space.ts
+++ b/node_modules/@koinos/sdk-as/assembly/util/space.ts
@@ -3,6 +3,8 @@ import { chain } from '@koinos/proto-as';
 import { Protobuf, Reader, Writer } from 'as-proto';
 import { StringBytes } from './stringBytes';
 
+const defaultKey = new Uint8Array(0);
+
 export namespace Space {
   export enum Direction {
     Ascending,
@@ -11,6 +13,7 @@ export namespace Space {
 
   export class Space<TKey, TValue> {
     private space: chain.object_space;
+    private defaultValue: TValue | null;
     private valueDecoder: (reader: Reader, length: i32) => TValue;
     private valueEncoder: (message: TValue, writer: Writer) => void;
 
@@ -18,6 +21,7 @@ export namespace Space {
     * Initialize a Space object with TKey the type of the keys and TValue the type of the values
     * @param contractId the id of the contract
     * @param spaceId the id of the space
+    * @param defaultValue default value
     * @param valueDecoder the protobuf decoder for the values
     * @param valueEncoder the protobuf encoder for the values
     * @param system is system space
@@ -31,10 +35,12 @@ export namespace Space {
     constructor(
       contractId: Uint8Array,
       spaceId: u32,
+      defaultValue: TValue | null,
       valueDecoder: (reader: Reader, length: i32) => TValue,
       valueEncoder: (message: TValue, writer: Writer) => void,
       system: bool = false) {
       this.space = new chain.object_space(system, contractId, spaceId);
+      this.defaultValue = defaultValue;
       this.valueDecoder = valueDecoder;
       this.valueEncoder = valueEncoder;
     }
@@ -71,7 +77,9 @@ export namespace Space {
     * ```
     */
     get(key: TKey): TValue | null {
-      return System.getObject<TKey, TValue>(this.space, key, this.valueDecoder);
+      const value = System.getObject<TKey, TValue>(this.space, key, this.valueDecoder);
+      if (!value && this.defaultValue) return this.defaultValue;
+      return value;
     }
 
     /**
@@ -483,4 +491,77 @@ export namespace Space {
       super.remove(finalKey);
     }
   }
+
+  export class Value<TValue> {
+    private space: chain.object_space;
+    private defaultValue: TValue | null;
+    private valueDecoder: (reader: Reader, length: i32) => TValue;
+    private valueEncoder: (message: TValue, writer: Writer) => void;
+
+    /**
+    * Initialize a Space object with TKey the type of the keys and TValue the type of the values
+    * @param contractId the id of the contract
+    * @param spaceId the id of the space
+    * @param defaultValue default value
+    * @param valueDecoder the protobuf decoder for the values
+    * @param valueEncoder the protobuf encoder for the values
+    * @param system is system space
+    * @example
+    * ```ts
+    * const contractId = Base58.decode('1DQzuCcTKacbs9GGScFTU1Hc8BsyARTPqe');
+    * const spaceId = 1;
+    * const obj = new Space.Value<test_object>(contractId, spaceId, test_object.decode, test_object.encode);
+    * ```
+    */
+    constructor(
+      contractId: Uint8Array,
+      spaceId: u32,
+      defaultValue: TValue | null,
+      valueDecoder: (reader: Reader, length: i32) => TValue,
+      valueEncoder: (message: TValue, writer: Writer) => void,
+      system: bool = false) {
+      this.space = new chain.object_space(system, contractId, spaceId);
+      this.defaultValue = defaultValue;
+      this.valueDecoder = valueDecoder;
+      this.valueEncoder = valueEncoder;
+    }
+
+    /**
+    * Get the object from the space
+    * @returns the object if exists, or the defaultValue
+    * if exists, null otherwise
+    * @example
+    * ```ts
+    * const myobj = obj.get();
+    * ```
+    */
+    get(): TValue | null {
+      const value = System.getObject<Uint8Array, TValue>(this.space, defaultKey, this.valueDecoder);
+      if (!value && this.defaultValue) return this.defaultValue;
+      return value;
+    }
+
+    /**
+    * Put an object in the space
+    * @param object object to put
+    * @example
+    * ```ts
+    * obj.put(new test_object(42));
+    * ```
+    */
+    put(object: TValue): void {
+      System.putObject(this.space, defaultKey, object, this.valueEncoder);
+    }
+
+    /**
+    * Remove the object from the space
+    * @example
+    * ```ts
+    * obj.remove();
+    * ```
+    */
+     remove(): void {
+      System.removeObject(this.space, defaultKey);
+    }
+  }
 }
